multiboot2/src/lib.rs

```diff
@@ -476,10 +494,62 @@ impl Reader {
     }
 }
 
+/// A trait to abstract over all sized and unsized tags (DSTs). For sized tags,
+/// this trait does not much. For DSTs, a `TagTrait::dst_size` implementation
+/// must me provided, which returns the right size hint for the dynamically
+/// sized portion of the struct.
+///
+/// The [`TagTrait::from_base_tag`] method has a default implementation for all
+/// tags that are `Sized`.
+///
+/// # Trivia
+/// This crate uses the [`Pointee`]-abstraction of the [`ptr_meta`] crate to
+/// create fat pointers.
+pub trait TagTrait: Pointee {
+    /// Returns
+    fn dst_size(base_tag: &Tag) -> Self::Metadata;
+
+    /// Creates a reference to a (dynamically sized) tag type in a safe way.
+    /// DST tags need to implement a proper [`Self::dst_size`] implementation.
+    ///
+    /// # Safety
+    /// Callers must be sure that the "size" field of the provided [`Tag`] is
+    /// sane and the underlying memory valid. The implementation of this trait
+    /// **must have** a correct [`Self::dst_size`] implementation.
+    unsafe fn from_base_tag<'a>(tag: &Tag) -> &'a Self {
+        let ptr = tag as *const _ as *const ();
+        let ptr = ptr_meta::from_raw_parts(ptr, Self::dst_size(tag));
+        &*ptr
+    }
+}
+
+// All sized tags automatically have a Pointee implementation where
+// Pointee::Metadata is (). Hence, the TagTrait is implemented automatically for
+// all tags that are sized.
+impl<T: Pointee<Metadata = ()>> TagTrait for T {
+    #[allow(clippy::unused_unit)]
+    fn dst_size(_: &Tag) -> Self::Metadata {
+        ()
+    }
+}
+
+// All sized tags automatically have a Pointee implementation where
+// Pointee::Metadata is (). Hence, the TagTrait is implemented automatically for
+// all tags that are sized.
+impl<T: Pointee<Metadata = usize>> TagTrait for T {
```

Comment:
hey @nicholasbishop,
May I ask you for help? That's a tough one, I think. I want to have default implementations for both cases but the compiler tells me the following

![image](https://github.com/rust-osdev/multiboot2/assets/5737016/0c304f1b-3b47-45d8-a5f6-499eb56871d8)

However, I'd assume that both implementations are distinct, as Metadata has a different type..

Reply:
Possibly the advice in https://stackoverflow.com/questions/40392524 would help, although I had some difficulty following it.

I'm not quite clear on whether a blanket impl makes sense here anyway though. Do I have it right that all tag types must start with a 32-bit ID followed by a 32-bit size field? The blanket impl would mean that `TagTrait` is implemented for things like `()` and `u32` which seems wrong.

Reply:
I had a chat with a colleque. After looking at <https://github.com/rust-lang/rust/issues/20400> and <https://github.com/rust-lang/rfcs/pull/1672#issuecomment-1405377983>, I came to the conclusion that the solution without <https://github.com/rust-osdev/multiboot2/pull/134/commits/4a6964ed09ca6ad3caf36c28d9863ea7ad1bf556> is good enough.